contrib/babelfishpg_tsql/sql/sys_functions.sql

```diff
@@ -3944,3 +3944,290 @@ RETURNS sys.SYSNAME AS $$
+            IF date_trunc('millisecond', result_time) > date THEN
+                RETURN result_time - concat(number, ' minutes')::interval;
+            END IF;
+            RETURN result_time;
+
+        WHEN 'second' THEN
+            required_bucket := ((hours_diff * 60 + minutes_diff) * 60 + seconds_diff)/number;
+            result_time := origin + concat(required_bucket * number, ' ', 'seconds')::interval;
+            IF date_trunc('millisecond', result_time) > date THEN
+                RETURN result_time - concat(number, ' ', 'seconds')::interval;
+            END IF;
+            RETURN result_time;
+
+        WHEN 'millisecond' THEN
+            required_bucket := (((hours_diff * 60 + minutes_diff) * 60) * 1000 + milliseconds_diff)/number;
+            result_time := origin + concat(required_bucket * number,' milliseconds')::interval;
+            IF date_trunc('millisecond', result_time) > date THEN
+                RETURN result_time - concat(number, ' milliseconds')::interval;
+            END IF;
+            RETURN result_time;
+        END CASE;
+
+    -- support of date_bucket() when date is of {'datetime2', 'datetimeoffset'} datatype
+    -- handling separately because both the datatypes have precision in milliseconds
+    ELSIF date_arg_datatype IN ('sys.datetime2'::regtype, 'sys.datetimeoffset'::regtype) THEN
+        -- when datepart is {year, quarter, month} make use of AGE() function to find number of buckets
+        IF datepart IN ('year', 'quarter', 'month') THEN
+            date_difference_interval := AGE(date_trunc('day', date::timestamp), date_trunc('day', origin::timestamp));
+            years_diff := EXTRACT('Year' from date_difference_interval)::INT;
+            months_diff := EXTRACT('Month' from date_difference_interval)::INT;
+            CASE datepart
+            WHEN 'year' THEN
+                -- Here we are finding how many buckets we have to add in the origin so that we can reach to a bucket in whcih our date falls.
+                -- For cases where origin > date, we might end up in a bucket which exceeds date by 1 bucket. (Ex. date_bucket(year, 2, '2010-01-01', '2019-01-01')) 
+                required_bucket := years_diff/number;
+                result_date := origin::timestamp + concat(required_bucket * number, ' years')::interval;
+                IF result_date > date::timestamp THEN
+                    result_date = result_date - concat(number, ' years')::interval;
+                END IF;
+
+            WHEN 'month' THEN
+                required_bucket := (12 * years_diff + months_diff)/number;
+                result_date := origin::timestamp + concat(required_bucket * number, ' months')::interval;
+                IF result_date > date::timestamp THEN
+                    result_date = result_date - concat(number, ' months')::interval;
+                END IF;
+
+            WHEN 'quarter' THEN
+                years_diff := (12 * years_diff + months_diff)/3;
```

Comment:
Use a different variable name to avoid confusion.

Reply:
Done